OpenRA.Game/Network/Connection.cs

```diff
@@ -325,35 +334,118 @@ void NetworkConnectionReceive()
 			}
 		}
 
+		void Ack(byte[] buf)
+		{
+			int frameReceived;
+			short framesToAck;
+			using (var reader = new BinaryReader(new MemoryStream(buf)))
+			{
+				frameReceived = reader.ReadInt32();
+				reader.ReadByte();
+				framesToAck = reader.ReadInt16();
+			}
+
+			var ms = new MemoryStream(4 + awaitingAckPackets.Take(framesToAck).Sum(i => i.Length));
+			ms.WriteArray(BitConverter.GetBytes(frameReceived));
+
+			for (var i = 0; i < framesToAck; i++)
+			{
+				byte[] queuedPacket = default;
+				if (awaitingAckPackets.Count > 0 && !awaitingAckPackets.TryDequeue(out queuedPacket))
+				{
+					// The dequeuing failed because of concurrency, so we retry
+					for (var c = 0; c < 5; c++)
+					{
+						if (awaitingAckPackets.TryDequeue(out queuedPacket))
+						{
+							break;
+						}
+					}
+				}
+
+				if (queuedPacket == default)
+				{
+					throw new InvalidOperationException("Received acks for unknown frames");
+				}
+
+				ms.WriteArray(queuedPacket);
+			}
+
+			AddPacket(new ReceivedPacket { FromClient = LocalClientId, Data = ms.GetBuffer() });
+		}
+
 		public override int LocalClientId { get { return clientId; } }
 		public override ConnectionState ConnectionState { get { return connectionState; } }
 
 		public override void SendSync(int frame, byte[] syncData)
 		{
-			var ms = new MemoryStream(4 + syncData.Length);
-			ms.WriteArray(BitConverter.GetBytes(frame));
-			ms.WriteArray(syncData);
-			queuedSyncPackets.Add(ms.GetBuffer());
+			using (var ms = new MemoryStream(4 + syncData.Length))
+			{
+				ms.WriteArray(BitConverter.GetBytes(frame));
+				ms.WriteArray(syncData);
+
+				queuedSyncPackets.Add(ms.GetBuffer());
+			}
 		}
 
-		protected override void Send(byte[] packet)
+		// Override send frame orders so we can hold them until ACK'ed
+		public override void Send(int frame, IEnumerable<byte[]> orders)
 		{
-			base.Send(packet);
+			var ordersLength = orders.Sum(i => i.Length);
+			var ms = new MemoryStream(8 + ordersLength);
 
-			try
+			// Always write data for old netcode
+			if (orders.Count() > 0 || !UseNewNetcode)
```

Comment:
Avoid LINQ. Walking a complete list.

Reply:
In this case it should be fine, the underlying structure is a list. But the parameter `orders` could be kept as a list or changed to a array.